### PR TITLE
Update `is_safetensors_compatible` check

### PR DESCRIPTION
**Pull request by DN6**
# What does this PR do?
The `is_safetensors_compatible` check hasn't been updated in a while and I think the assumptions there are a bit dated. Basically the current version assumes that if a `.bin` file exists in the model repo, an equivalently named safetensors file must also exist for the check to pass. e.g if `model.bin` is present in the repo then `model.safetensors` must also be present for the check to pass. 

It doesn't account for cases where the safetensors file might be a variant of the `model.bin` file. e.g. if `model.bin` and only `model.fp16.safetensors` is present in a model folder the check will return False. 

This PR updates the check so that it returns True if a `safetensors` file is present for every component of a pipeline. 

One thing to note, is that this check will also return `True` if a repo has a mix of variants and non-variant files and `variant=None`

e.g
If the VAE only has fp16 variant and UNet only has the non-variant, the check still passes because the repo does contain safetensors files for all the components. 

```
"vae/diffusion_pytorch_model.fp16.safetensors",
"unet/diffusion_pytorch_model.safetensors",
```
 
This is probably an edge case though. And I think it's okay that the check returns True. It's purpose is to verify that safetensors files exist for all components and to check that the variant safetensors exists if the `variant` argument is provided. 

The pipeline will error out in `_get_model_file` in such a case anyway.         
      

Fixes # (issue)
#8968

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/diffusers/pipelines/pipeline_loading_utils.py [hunk 1]
```diff
@@ -101,37 +101,33 @@ def is_safetensors_compatible(filenames, variant=None, passed_components=None) -
     - For models from the transformers library, the filename changes from "pytorch_model" to "model", and the ".bin"
       extension is replaced with ".safetensors"
     """
-    pt_filenames = []
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/pipeline_loading_utils.py:104]: maybe we can use `model_folder_names` directly to simplifies a little bit https://github.com/huggingface/diffusers/blob/73acebb8cfbd1d2954cabe1af4185f9994e61917/src/diffusers/pipelines/pipeline_utils.py#L1357

File: src/diffusers/pipelines/pipeline_loading_utils.py [hunk 2]
```diff
@@ -101,37 +101,33 @@ def is_safetensors_compatible(filenames, variant=None, passed_components=None) -
...
+    for component, component_filenames in components.items():
+        matches = []
+        for component_filename in component_filenames:
+            filename, extension = os.path.splitext(component_filename)
 
-        if filename.startswith("pytorch_model"):
-            filename = filename.replace("pytorch_model", "model")
-        else:
-            filename = filename
+            match_exists = (variant in filename if variant else True) and extension == ".safetensors"
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/pipeline_loading_utils.py:127]: would this work?
```suggestion
            if variant:
                match_exists = variant in filename and extension == ".safetensors"
             else:
                match_exists = variant not in filename and extension == ".safetensors"
```

**Review by yiyixuxu — COMMENTED** [review]
nice! yeah agree that this check does not make sense anymore

for this, I think we should be checking only non-variant files when `variant=None`, no? 

>One thing to note, is that this check will also return True if a repo has a mix of variants and non-variant files and variant=None

**Comment by DN6** [comment]
Okay. Turns out the check can be made much simpler since we support loading models if there are a mixture of variant/non-variant files. e.g. Some models have variants while others don't. 

So the only requirement for this check is to see if there is a `safetensors` file for each component. I do think the warning messages/checks for when we do mixed file loading could be a bit better, but that's probably better done in a separate PR.  

I also broke up one of the tests into 4 separate tests to make it a bit easier to debug failure cases.

**Review by yiyixuxu — APPROVED** [review]
thanks!

**Pull request merged**